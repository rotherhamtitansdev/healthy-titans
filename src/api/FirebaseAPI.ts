import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
