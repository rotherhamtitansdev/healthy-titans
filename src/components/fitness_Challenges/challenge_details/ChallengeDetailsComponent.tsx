import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import ChallengeDetailsComponentData from "../../../data/ChallengeDetailsComponentData";
import { ChallengeDetailsProps } from "../../../models/ChallengeDetailsProps";
import Card from "../../shared/Card";
import DetailsCard from "../../shared/DetailsCard";
import DetailsComponent from "../../shared/DetailsComponent";

const ChallengeDetailsComponent = () => {
  // eslint-disable-next-line max-len
  const [getChallengeDetailsComponentData, setChallengeDetailsComponentData] = useState<
    ChallengeDetailsProps | undefined
  >();

  const { challengeName } = useParams();
  useEffect(() => {
    setChallengeDetailsComponentData(
      ChallengeDetailsComponentData[
        challengeName as unknown as keyof typeof ChallengeDetailsComponentData
      ]
    );
  });

  return (
    <div>
      {getChallengeDetailsComponentData && (
        <DetailsComponent>
          <div className="grid grid-cols-2 gap-4">
            <DetailsCard
              name={getChallengeDetailsComponentData.name}
              description={getChallengeDetailsComponentData.description}
              img={getChallengeDetailsComponentData.img}
            />
            <Card
              card={{
                name: getChallengeDetailsComponentData.name,
                additionalStyling: "h-content"
              }}
            >
              <div className="pl-4 pr-4">
                <div className="pt-4 pb-3 tracking-wide text-lg text-black font-bold">Rules</div>
                <p className="pl-4 block mt-1 text-lg leading-tight font-semibold text-gray-900">
                  <ol className="list-decimal">
                    {getChallengeDetailsComponentData.rules.map((rule) => (
                      <li key={rule} className="pb-4">
                        {rule}
                      </li>
                    ))}
                  </ol>
                </p>
              </div>
            </Card>
          </div>
        </DetailsComponent>
      )}
    </div>
  );
};

export default ChallengeDetailsComponent;
