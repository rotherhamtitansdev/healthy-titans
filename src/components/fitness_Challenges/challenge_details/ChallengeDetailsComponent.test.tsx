import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router";
import FirebaseAPI from "../../../api/FirebaseAPI";
import ChallengeDetailsComponent from "./ChallengeDetailsComponent";

describe("ChallengeDetailsComponent", () => {
  test("component renders rules if in data", async () => {
    const content = {
      name: "Test Activity",
      description: "Sample description of the test activity",
      firebaseName: "firebase",
      rules: ["Rule 1", "Rule 2"],
      equipment: ["Ball", "Bow and Arrows"],
      howToPlay: ["Throw the ball", "Shoot the ball with the bow and arrrows"],
    };
    const url = "/test.jpg";

    const mockFetchChallengesData = jest.fn().mockResolvedValueOnce(content);

    jest.spyOn(FirebaseAPI, "fetchChallengesData").mockImplementation(mockFetchChallengesData);
    jest.spyOn(FirebaseAPI, "fetchImages").mockResolvedValue(url);

    const { asFragment } = render(
      <MemoryRouter initialEntries={["/FitnessChallenges/mychallenge"]} initialIndex={0}>
        <Routes>
          <Route path="/FitnessChallenges/:challengeName" element={<ChallengeDetailsComponent />} />
        </Routes>
      </MemoryRouter>
    );
    
    expect(mockFetchChallengesData).toBeCalledWith("mychallenge");
    await waitFor(() => {
      expect(screen.getByText("Ball")).toBeVisible();
    });
    
    expect(screen.getByText("Rules")).toBeVisible();

    expect(asFragment()).toMatchSnapshot();
  });


});
