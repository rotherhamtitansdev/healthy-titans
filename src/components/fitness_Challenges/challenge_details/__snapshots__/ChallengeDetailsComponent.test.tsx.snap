// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeDetailsComponent component renders rules if in data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="px-10 lg:px-6 pt-4 lg:container mx-auto"
    >
      <div
        class="text-[#243A73] flex font-bold sm:text-[18px] pb-10"
        data-testid="breadcrumbs"
      >
        <a
          aria-current="page"
          class="active"
          href="/FitnessChallenges"
        >
           Fitness Challenges
        </a>
        <p
          class="text text-titansBrightPink"
        >
           &gt; 
        </p>
        <a
          aria-current="page"
          class="active"
          href="/FitnessChallenges/mychallenge"
        >
           Mychallenge
        </a>
      </div>
      <div
        class="bg-white rounded-2xl flex flex-wrap md:flex-nowrap gap-y-5 gap-x-0 md:gap-x-2 mt-[6rem] mb-[6rem]"
      >
        <div
          class="pl-5 md:basis-1/2"
          data-testid="Test Activity-details"
        >
          <img
            alt=""
            class="absolute visible top-[13rem]"
            src="/test.jpg"
          />
          <div
            class="pl-2 pr-4 pt-[5rem]"
          >
            <div
              class="tracking-wide text-4xl text-black font-bold py-5 text-homepageHeaderSubTitle"
              data-testid="details-title"
            >
              Test Activity
            </div>
            <p
              class="block mt-1 text-xl leading-tight font-medium text-gray-900 text-[16px] lg:text-[20px]"
              data-testid="details-description"
            >
              Sample description of the test activity
            </p>
            <div
              class="block mt-1 text-lg leading-tight font-semibold text-gray-900"
            >
              <h2
                class="py-5 text-[20px] lg:text-[24px] "
              >
                Equipment
              </h2>
              <ul
                class="list-none px-0 lg:px-5 text-[16px] lg:text-[20px]"
              >
                <li
                  class="pb-4 md:pb-8 flex font-medium"
                >
                  <img
                    alt="checkmark"
                    class="pr-5 align-center"
                    src="CheckMarkButton.png"
                  />
                  Ball
                </li>
                <li
                  class="pb-4 md:pb-8 flex font-medium"
                >
                  <img
                    alt="checkmark"
                    class="pr-5 align-center"
                    src="CheckMarkButton.png"
                  />
                  Bow and Arrows
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="md:basis-1/2 "
        >
          <div
            class="pr-[2rem] md:pr-[3rem] lg:pr-[5rem] pt-[2rem]"
          >
            <div
              class="lg:pt-4 pb-3 lg:pl-0 pl-5 tracking-wide text-[20px] lg:text-[24px] font-quicksand text-homepageHeaderSubTitle font-semibold pb-[2rem]"
            >
              How to play
            </div>
            <div
              class="pl-4 block mt-1 text-lg leading-tight font-medium font-quicksand"
            >
              <ul
                class="list-decimal"
              >
                <li
                  class="pb-6"
                >
                  Throw the ball
                </li>
                <li
                  class="pb-6"
                >
                  Shoot the ball with the bow and arrrows
                </li>
              </ul>
            </div>
            <div
              class="pt-4 pb-3 tracking-wide text-2xl font-quicksand text-homepageHeaderSubTitle font-semibold"
            >
              Rules
            </div>
            <div
              class="pl-4 block mt-1 text-lg leading-tight font-medium font-quicksand"
            >
              <ul
                class="list-disc"
              >
                <li
                  class="pb-6"
                >
                  Rule 1
                </li>
                <li
                  class="pb-6"
                >
                  Rule 2
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
